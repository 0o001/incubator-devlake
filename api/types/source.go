package types

type CreateSource struct {
	Plugin  string                 `json:"plugin" binding:"required"`
	Name    string                 `json:"name"`
	Options map[string]interface{} `json:"options" binding:"required"`
}
