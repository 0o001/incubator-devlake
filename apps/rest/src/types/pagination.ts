import { Type } from 'class-transformer';
import { Max, Min } from 'class-validator';

export class PaginationRequest {
  @Min(1)
  @Type(() => Number)
  page: number;

  @Max(100)
  @Type(() => Number)
  pagesize: number;
}

export class PaginationResponse<T> {
  total: number;
  offset: number;
  page: number;
  pagesize: number;
  data: T[];
}
