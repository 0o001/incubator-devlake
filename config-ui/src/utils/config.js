export const DEVLAKE_ENDPOINT = '/api'
export const GRAFANA_BASE_URL = '/grafana'
export const LOCAL_BASE_URL = 'http://localhost:3002'
export const GRAFANA_ENDPOINT = '/d/0Rjxknc7z/demo-homepage?orgId=1'
