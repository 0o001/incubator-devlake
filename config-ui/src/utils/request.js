import axios from 'axios'

const headers = {}

export default {
  post: async (url, body) => {
    return await axios.post(
      url,
      body,
      {
        headers
      }
    )
  },
  get: async (url) => {
    return await axios.get(
      url,
      {
        headers
      }
    )
  }
}
