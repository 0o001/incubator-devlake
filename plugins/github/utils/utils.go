package utils

import (
	"errors"
	"fmt"
	"net/http"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/merico-dev/lake/logger"
)

type PagingInfo struct {
	Next  int
	Last  int
	First int
	Prev  int
}

type RateLimitInfo struct {
	Date      time.Time
	ResetTime time.Time
	Remaining int
}

func ConvertRateLimitInfo(date string, resetTime string, remaining string) (RateLimitInfo, error) {
	convertedDate, err := http.ParseTime(date)
	if err != nil {
		fmt.Println("Convert error: ", err)
	}
	resetInt, err := strconv.ParseInt(resetTime, 10, 64)
	if err != nil {
		fmt.Println("Convert error: ", err)
	}
	convertedResetTime := time.Unix(resetInt, 0)
	convertedRemaining, err := strconv.Atoi(remaining)
	if err != nil {
		logger.Error("Convert error: ", err)
	}
	return RateLimitInfo{
		Date:      convertedDate,
		ResetTime: convertedResetTime,
		Remaining: convertedRemaining,
	}, nil
}

func GetRateLimitPerSecond(info RateLimitInfo) int {
	return info.Remaining / int(info.ResetTime.Unix()-info.Date.Unix()) * 9 / 10
}
func ConvertStringToInt(input string) (int, error) {
	return strconv.Atoi(input)
}
func GetPagingFromLinkHeader(link string) (PagingInfo, error) {
	result := PagingInfo{
		Next:  0,
		Last:  0,
		Prev:  0,
		First: 0,
	}
	linksArray := strings.Split(link, ",")
	pattern1 := regexp.MustCompile(`page=*[0-9]+`)
	pattern2 := regexp.MustCompile(`rel="*[a-z]+`)
	if len(linksArray) >= 2 {
		for i := 0; i < len(linksArray); i++ {
			content := []byte(linksArray[i])
			loc1 := pattern1.FindIndex(content)
			loc2 := pattern2.FindIndex(content)
			if len(loc1) >= 2 && len(loc2) >= 2 {
				pageNumberSubstring := string(content[loc1[0]:loc1[1]])
				pageNumberString := strings.Replace(pageNumberSubstring, `page=`, ``, 1)
				pageNameSubstring := string(content[loc2[0]:loc2[1]])
				pageNameString := strings.Replace(pageNameSubstring, `rel="`, ``, 1)

				pageNumberInt, convertErr := ConvertStringToInt(pageNumberString)
				if convertErr != nil {
					return result, convertErr
				}
				switch pageNameString {
				case "next":
					result.Next = pageNumberInt

				case "first":
					result.First = pageNumberInt

				case "last":
					result.Last = pageNumberInt

				case "prev":
					result.Prev = pageNumberInt
				}

			} else {
				return result, errors.New("Parsed string values aren't long enough.")
			}
		}
		return result, nil
	} else {
		return result, errors.New("The link string provided is invalid.")
	}
}
