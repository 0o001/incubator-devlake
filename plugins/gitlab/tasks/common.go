package tasks

import (
	"fmt"
	"time"

	"github.com/merico-dev/lake/config"
	"github.com/merico-dev/lake/plugins/core"
)

func CreateApiClient() *core.ApiClient {
	return core.NewApiClient(
		config.V.GetString("GITLAB_ENDPOINT"),
		map[string]string{
			"Authorization": fmt.Sprintf("Bearer %v", config.V.GetString("GITLAB_AUTH")),
		},
		10*time.Second,
		3,
	)
}
