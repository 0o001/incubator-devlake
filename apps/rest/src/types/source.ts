import { IsNotEmpty, IsOptional } from 'class-validator';
import { PaginationRequest } from './pagination';

export class CreateSource {
  @IsNotEmpty()
  type: string;

  @IsOptional()
  name?: string;

  @IsNotEmpty()
  options: Record<string, unknown>;
}

export class UpdateSource extends CreateSource {}

export class ListSource extends PaginationRequest {
  @IsOptional()
  type?: string;
}
