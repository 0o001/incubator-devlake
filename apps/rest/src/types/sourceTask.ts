import { IsArray, IsNotEmpty, IsOptional } from 'class-validator';
import { PaginationRequest } from './pagination';

export class CreateSourceTask {
  @IsArray()
  collector: string[];

  @IsArray()
  enricher: string[];

  @IsNotEmpty()
  options: Record<string, unknown>;
}

export class ListSourceTask extends PaginationRequest {
  @IsOptional()
  source_id?: string;
}
