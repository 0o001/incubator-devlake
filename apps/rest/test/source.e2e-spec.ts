import { Test, TestingModule } from '@nestjs/testing';
import { HttpStatus, INestApplication } from '@nestjs/common';
import * as request from 'supertest';
import { AppModule } from '../src/app.module';
import { ignoreEntityProps, truncateTableForTest } from './utils';
import * as uuid from 'uuid';
import { PaginationResponse } from '../src/types/pagination';
import Source from '../src/models/source';
import * as _ from 'lodash';

describe('SourceController (e2e)', () => {
  let app: INestApplication;

  beforeAll(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();
  });

  afterAll(async () => {
    await app.close();
  });

  beforeEach(async () => {
    await truncateTableForTest(['source']);
  });

  describe('/source (POST)', () => {
    it('should return source type', () => {
      const newSource = {
        type: 'jira',
        options: {
          host: 'https://www.atlassian.com/',
          email: 'xx@example.com',
          auth: 'base64EncodedAuthToken',
        },
      };
      return request(app.getHttpServer())
        .post('/source')
        .send(newSource)
        .expect(201)
        .expect((res) => {
          expect(ignoreEntityProps(res.body)).toMatchSnapshot();
          expect(uuid.validate(res.body.id)).toBe(true);
        });
    });

    it('should return validate error', () => {
      const sourceWithoutType = {
        options: {},
      };
      return request(app.getHttpServer())
        .post('/source')
        .send(sourceWithoutType)
        .expect(400)
        .expect((res) => {
          expect(res.body).toMatchSnapshot();
        });
    });

    it('should return validate error 2', () => {
      const sourceWithoutType = {
        type: 'github',
        options: {},
      };
      return request(app.getHttpServer())
        .post('/source')
        .send(sourceWithoutType)
        .expect(400)
        .expect((res) => {
          expect(res.body).toMatchSnapshot();
        });
    });

    it('should return validate error 3', () => {
      const sourceWithoutType = {
        type: 'gitlab',
      };
      return request(app.getHttpServer())
        .post('/source')
        .send(sourceWithoutType)
        .expect(400)
        .expect((res) => {
          expect(res.body).toMatchSnapshot();
        });
    });
  });

  describe('/source/:id (GET)', () => {
    it('should return not found if record not exists', async () => {
      return request(app.getHttpServer())
        .get('/source/id-not-exist')
        .expect(404)
        .expect((res) => {
          expect(res.body).toMatchSnapshot();
        });
    });
  });

  describe('/source (CRUD)', () => {
    it('mock a CRUD progress', async () => {
      const server = request(app.getHttpServer());
      // create a jira source
      const jiraSource = {
        type: 'jira',
        options: {
          host: 'https://www.atlassian.com/',
          username: 'lake@merico.com',
          token: 'guess what?',
        },
      };
      let response = await server.post('/source').send(jiraSource);
      expect(response.status).toBe(HttpStatus.CREATED);

      const jira = response.body;
      expect(ignoreEntityProps(jira)).toMatchSnapshot();

      //  create a gitlab source
      const gitlabSource = {
        type: 'gitlab',
        options: {
          host: 'https://www.gitlab.com',
          username: 'lake@merico.com',
          token: 'gitlab token',
        },
      };

      response = await server.post('/source').send(gitlabSource);
      expect(response.status).toBe(HttpStatus.CREATED);

      const gitlab = response.body;
      expect(ignoreEntityProps(gitlab)).toMatchSnapshot();

      // list sources
      response = await server.get('/source').query({
        page: 1,
        pagesize: 10,
      });

      expect(response.status).toBe(HttpStatus.OK);
      let data: PaginationResponse<Source> = response.body;
      expect(_.omit(data, 'data')).toMatchSnapshot();
      expect(
        data.data
          .map((item) => ignoreEntityProps(item))
          .sort((a, b) => (a.type < b.type ? 1 : -1)),
      ).toMatchSnapshot();

      // filter sources
      response = await server.get('/source').query({
        page: 1,
        pagesize: 10,
        type: 'jira',
      });
      data = response.body;
      expect(_.omit(data, 'data')).toMatchSnapshot();
      expect(
        data.data.map((item) => ignoreEntityProps(item)),
      ).toMatchSnapshot();

      // update source
      await server.put(`/source/${gitlab.id}`).send({
        options: {
          token: 'updated token',
        },
      });
      // get source
      response = await server.get(`/source/${gitlab.id}`);
      expect(ignoreEntityProps(response.body)).toMatchSnapshot();

      // delete source
      await server.delete(`/source/${gitlab.id}`);

      response = await server.get(`/source/${gitlab.id}`);
      expect(response.status).toEqual(HttpStatus.NOT_FOUND);
    });
  });
});
