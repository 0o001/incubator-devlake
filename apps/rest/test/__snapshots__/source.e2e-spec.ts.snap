// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SourceController (e2e) /source (CRUD) mock a CRUD progress 1`] = `
Object {
  "name": "cloud",
  "options": Object {
    "host": "https://www.atlassian.com/",
    "token": "guess what?",
    "username": "lake@merico.com",
  },
  "type": "jira",
}
`;

exports[`SourceController (e2e) /source (CRUD) mock a CRUD progress 2`] = `
Object {
  "name": null,
  "options": Object {
    "host": "https://www.gitlab.com",
    "token": "gitlab token",
    "username": "lake@merico.com",
  },
  "type": "gitlab",
}
`;

exports[`SourceController (e2e) /source (CRUD) mock a CRUD progress 3`] = `
Object {
  "offset": 0,
  "page": 1,
}
`;

exports[`SourceController (e2e) /source (CRUD) mock a CRUD progress 4`] = `
Array [
  Object {
    "name": "cloud",
    "options": Object {
      "host": "https://www.atlassian.com/",
      "token": "guess what?",
      "username": "lake@merico.com",
    },
    "type": "jira",
  },
  Object {
    "name": null,
    "options": Object {
      "host": "https://www.gitlab.com",
      "token": "gitlab token",
      "username": "lake@merico.com",
    },
    "type": "gitlab",
  },
]
`;

exports[`SourceController (e2e) /source (CRUD) mock a CRUD progress 5`] = `
Object {
  "offset": 0,
  "page": 1,
}
`;

exports[`SourceController (e2e) /source (CRUD) mock a CRUD progress 6`] = `
Array [
  Object {
    "name": "cloud",
    "options": Object {
      "host": "https://www.atlassian.com/",
      "token": "guess what?",
      "username": "lake@merico.com",
    },
    "type": "jira",
  },
]
`;

exports[`SourceController (e2e) /source (CRUD) mock a CRUD progress 7`] = `
Object {
  "name": null,
  "options": Object {
    "host": "https://www.gitlab.com",
    "token": "gitlab token",
    "username": "lake@merico.com",
  },
  "type": "gitlab",
}
`;

exports[`SourceController (e2e) /source (POST) should return source type 1`] = `
Object {
  "name": "cloud",
  "options": Object {
    "auth": "base64EncodedAuthToken",
    "email": "xx@example.com",
    "host": "https://www.atlassian.com/",
  },
  "type": "jira",
}
`;

exports[`SourceController (e2e) /source (POST) should return validate error 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    "type should not be empty",
  ],
  "statusCode": 400,
}
`;

exports[`SourceController (e2e) /source (POST) should return validate error 3 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    "options should not be empty",
  ],
  "statusCode": 400,
}
`;

exports[`SourceController (e2e) /source/:id (GET) should return not found if record not exists 1`] = `
Object {
  "message": "Could not find any entity of type \\"Source\\" matching: \\"id-not-exist\\"",
}
`;
